elevation_layer/src/elevation_layer.cpp

```diff
@@ -0,0 +1,236 @@
+    double px = vertexPositionXY.x();
+    double py = vertexPositionXY.y();
+
+    touch(px, py, min_x, min_y, max_x, max_y);
+  }
+  updateFootprint(robot_x, robot_y, robot_yaw, min_x, min_y, max_x, max_y);
+}
+
+void ElevationLayer::updateFootprint(double robot_x, double robot_y, double robot_yaw, double* min_x, double* min_y, double* max_x,
+                                     double* max_y) {
+  if (!footprint_clearing_enabled_) return;
+  costmap_2d::transformFootprint(robot_x, robot_y, robot_yaw, getFootprint(), transformed_footprint_);
+
+  for (auto& i : transformed_footprint_) {
+    touch(i.x, i.y, min_x, min_y, max_x, max_y);
+  }
+}
+
+void ElevationLayer::updateCosts(costmap_2d::Costmap2D& master_grid, int min_i, int min_j, int max_i, int max_j) {
+  std::lock_guard<std::mutex> lock(elevation_map_mutex_);
+  if (!enabled_ || !elevation_map_received_) return;
+  const bool has_edges_layer = elevation_map_.exists(edges_layer_name_);
+  if (!has_edges_layer) {
+    ROS_WARN_THROTTLE(0.2, "No edges layer found !!");
+  }
+  ros::Duration time_since_elevation_map_received = ros::Time::now() - last_elevation_map_update_;
+  if (time_since_elevation_map_received > ros::Duration(max_allowed_blind_time_)) {
+    current_ = false;
+  }
+  const grid_map::Matrix& elevation_data = elevation_map_[elevation_layer_name_];
+  for (grid_map::GridMapIterator iterator(elevation_map_); !iterator.isPastEnd(); ++iterator) {
+    const grid_map::Index gridmap_index(*iterator);
+    grid_map::Position vertexPositionXY;
+    elevation_map_.getPosition(gridmap_index, vertexPositionXY);
+    double px = vertexPositionXY.x();
+    double py = vertexPositionXY.y();
+    // now we need to compute the map coordinates for the observation
+    unsigned int mx, my;
+    if (!worldToMap(px, py, mx, my))  // if point outside of local costmap, ignore
+    {
+      continue;
+    }
+    if (elevation_data(gridmap_index(0), gridmap_index(1)) > height_threshold_)  // If point too high, it could be an obstacle
+    {
+      if (has_edges_layer) {
+        const grid_map::Matrix& edges_data = elevation_map_[edges_layer_name_];
+        if (edges_data(gridmap_index(0), gridmap_index(1)) < edges_sharpness_threshold_)  // if area not sharp, dont label as obstacle
+        {
+          setCost(mx, my, FREE_SPACE);
+          continue;
+        }
+      }
+      setCost(mx, my, LETHAL_OBSTACLE);
+    } else {
+      setCost(mx, my, FREE_SPACE);
+    }
+  }
+
+  if (footprint_clearing_enabled_) {
+    setConvexPolygonCost(transformed_footprint_, costmap_2d::FREE_SPACE);
+  }
+
+  switch (combination_method_) {
+    case Overwrite:
+      updateWithOverwrite(master_grid, min_i, min_j, max_i, max_j);
+      break;
+    case Maximum:
+      updateWithMax(master_grid, min_i, min_j, max_i, max_j);
+      break;
+    default:  // Do Nothing
+      break;
+  }
+}
+
+CombinationMethod convertCombinationMethod(const std::string& str) {
+  if (str == "Overwrite") {
```

Comment:
You could use https://en.cppreference.com/w/cpp/string/byte/tolower to make it case insensitive...